mattermost/platform
# System Console: Email notification content setting

#### Summary
Added a new system console setting called Email Notification Content Type. This setting allows system administrators to suppress post contents in email notifications that are sent from secure systems. It can only be changed on licensed E20 instances. The default setting, "full", includes message contents in notification emails. The optional setting, "generic", removes message contents and other sensitive information from notification emails. This setting affects batched and non-batched emails. I also tried to do some refactoring around the way that we assemble emails to improve testability.

**Test steps:**
* The easiest way to test this is probably by checking out the branch in a local dev environment where you can access the Inbucket mock email server
* Add a license file to your instance
* Connect to your local database, find the UserId of your primary user account in the Users table, and change the email_interval setting for that user id in the Preferences table to 60. 
  * This will cause batched emails to be sent after 60s instead of the default 30s. 
  * If you use the default value, emails will be sent immediately, and the batching logic will be skipped.
* In System Console > Notifications > Email, activate Email Batching, and set Email Notification Contents to "full"
  * Use a separate user account in an incognito window to send public, private, and direct messages to your primary user account within a 1 minute time span
  * Check the format of the email received by the primary user account in Inbucket
  * It should include the full text of all three messages that were sent
* In system Console > Notifications > Email, activate Email Batching, and set Email Notification Contents to "generic"
  * Use a separate user account in an incognito window to send public, private, and direct messages to your primary user account within a 1 minute time span
  * Check the format of the email received by the primary user account in Inbucket
  * Each of the three messages should be listed, but the message text should **not** be included in the email
* In system Console > Notifications > Email, deactivate Email Batching, and set Email Notification Contents to "full"
  * Use a separate user account in an incognito window to send public, private, and direct messages to your primary user account
  * Check the format of each of the three emails received by the primary user account in Inbucket
  * Each of the three emails should include the message text and channel details
* In system Console > Notifications > Email, deactivate Email Batching, and set Email Notification Contents to "generic"
  * Use a separate user account in an incognito window to send public, private, and direct messages to your primary user account
  * Check the format of each of the three emails received by the primary user account in Inbucket
  * Each of the three emails should **not** include the message text or the channel details

**NOTE:** I'm still working on unit tests for `app/notifications.go`, but I wanted to get this PR up sooner than later, as this feature is intended for feature cut off today

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-7195

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Has enterprise changes (please link)
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
